### PR TITLE
Instrumenting a few commands

### PR DESCRIPTION
**Pull request by ellensi**
### Purpose

Adding instrumentation to a few Dynamo commands. Unlike the [previous PR](https://github.com/Conceptual-Design/Dynamo/pull/4), the event tracking is not centralized before/after the DelegateCommand execution. Instead, the events are only tracked when the actual user action is done (please refer to the rightmost column in the tables).

#### File/node operations

Category | Action | Sample description | Sample value | When?
--- | --- | --- | --- | --- | --- | ---
FileOperation | FileOpen Open | - | - | After file is opened
NodeOperations | Unresolved | `Translation.dll:Dynamo.Translation.FileLoader.FromPath@string, :Dynamo.Translation.Models.ObjectSelector, Translation.dll:Dynamo.Translation.ImportedObject .ConvertToGeometries@Dynamo.Translation.ImportedObject[]` | `3` | Only if there is any unresolved node
NodeOperations | Create | `Office.ReadFromFile` | - | After node is created

#### Commands

Category | Action | Description | Value | When?
--- | --- | --- | --- | ---
Command | ExportToSTL ~~Start~~ | - | - | STL file export location is specified and user confirms OK
Command | GraphLayout ~~Start~~ | - | - | All (menu, context menu, or Ctrl+L)
Command | NewCustomNode ~~Start~~ | `NodeCount` | Number of selected nodes | A new custom node workspace has been added
Command | ImageCapture ~~Start~~ | `NodeCount` | Number of all nodes in the workspace | Image export location is specified and user confirms OK
Command | Undo ~~Start~~ | - | - | Workspace is undone
Command | Redo ~~Start~~ | - | - | Workspace is redone

### Reviewer

@sharadkjaiswal

### CODE REVIEW

File: src/DynamoCore/Models/RecordableCommands.cs [hunk 1]
```diff
@@ -504,6 +504,17 @@ internal static OpenFileCommand DeserializeCore(XmlElement element)
...
                 dynamoModel.OpenFileImpl(this);
+
+                // Log file open action and the number of nodes in the opened workspace
+                Dynamo.Logging.Analytics.CreateFileOperationEvent(
+                    XmlFilePath, Logging.Actions.Open, dynamoModel.CurrentWorkspace.Nodes.Count());
+
+                // Log each unresolved node name in the opened workspace
+                foreach (DummyNode n in dynamoModel.CurrentWorkspace.Nodes.Where(x => x is DummyNode))
+                {
+                    Dynamo.Logging.Analytics.TrackEvent(Logging.Actions.Open, Logging.Categories.NodeOperations,
```
Comment (sharadkjaiswal) [inline src/DynamoCore/Models/RecordableCommands.cs]: Can we track this in one event, with a new Action `Unresolved`, value as number of unresolved, Description as comma separated string?

File: src/DynamoCore/Models/RecordableCommands.cs [hunk 2]
```diff
@@ -504,6 +504,17 @@ internal static OpenFileCommand DeserializeCore(XmlElement element)
             protected override void ExecuteCore(DynamoModel dynamoModel)
             {
                 dynamoModel.OpenFileImpl(this);
+
+                // Log file open action and the number of nodes in the opened workspace
+                Dynamo.Logging.Analytics.CreateFileOperationEvent(
```
Comment (sharadkjaiswal) [inline src/DynamoCore/Models/RecordableCommands.cs]: Should we rename this method as TrackFileOperationEvent and still return IDisposable?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 3]
```diff
@@ -1786,6 +1786,9 @@ internal bool CanDelete(object parameters)
         public void SaveImage(object parameters)
         {
             OnRequestSaveImage(this, new ImageSaveEventArgs(parameters.ToString()));
+
+            Dynamo.Logging.Analytics.CreateCommandEvent("SaveImageCapture",
```
Comment (sharadkjaiswal) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: ImageCapture or SaveImage is ok.

File: src/DynamoCoreWpf/ViewModels/Core/WorkspaceViewModel.cs [hunk 4]
```diff
@@ -1071,6 +1073,9 @@ internal void CollapseSelectedNodes()
             DynamoViewModel.Model.AddCustomNodeWorkspace(
                 DynamoViewModel.Model.CustomNodeManager.Collapse(selectedNodes,
                 selectedNotes, Model, DynamoModel.IsTestMode, args));
+
+            Dynamo.Logging.Analytics.CreateCommandEvent("NewCustomNode",
```
Comment (sharadkjaiswal) [inline src/DynamoCoreWpf/ViewModels/Core/WorkspaceViewModel.cs]: We can also include Node count as value.

File: src/DynamoCore/Models/RecordableCommands.cs [hunk 5]
```diff
@@ -504,6 +504,20 @@ internal static OpenFileCommand DeserializeCore(XmlElement element)
...
+                    XmlFilePath, Logging.Actions.Open, dynamoModel.CurrentWorkspace.Nodes.Count());
+
+                // Log unresolved node names and node count in the opened workspace
+                var unresolvedNodes = dynamoModel.CurrentWorkspace.Nodes.OfType<DummyNode>();
+                if (unresolvedNodes != null && unresolvedNodes.Any())
+                {
+                    Dynamo.Logging.Analytics.TrackEvent(Logging.Actions.Unresolved, Logging.Categories.NodeOperations,
```
Comment (lukechurch) [inline src/DynamoCore/Models/RecordableCommands.cs]: This needs to go to instrumentation not analytics as the assembly could contain names.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 6]
```diff
@@ -1791,6 +1791,9 @@ internal bool CanDelete(object parameters)
...
+                "Nodes", CurrentSpace.Nodes.Count());
```
Comment (lukechurch) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Consider "NodeCounts" instead of "Nodes" to make it clear that it's just a number

**Comment by sharadkjaiswal** [comment]
Thanks @ellensi this looks great. I was wondering if we should keep the action for command event as command name instead of command name + Start.

File: src/DynamoCore/Models/RecordableCommands.cs [hunk 7]
```diff
@@ -768,6 +797,14 @@ protected override void SerializeCore(XmlElement element)
                 }
             }
 
+            internal override void TrackAnalytics()
+            {
+                Dynamo.Logging.Analytics.TrackEvent(
+                    Logging.Actions.Create,
+                    Logging.Categories.NodeOperations,
+                    Node.Name);
```
Comment (sharadkjaiswal) [inline src/DynamoCore/Models/RecordableCommands.cs]: I am not sure if Node.Name is not always populated and has value. You can either use NickName or the extension method OriginalName().

**Comment by sharadkjaiswal** [comment]
please merge this change to RC branch.

**Pull request merged**